platforms/core-runtime/INSTRUMENTATION.md

```diff
@@ -0,0 +1,437 @@
+	    JavaCompileInterceptorsDeclaration.intercept_setSourceCompatibility(task, "1.8");
+     });
+}
+```
+
+Where JavaCompileInterceptorsDeclaration is a class defined by Gradle developer to replace original logic and the replacement is done via bytecode instrumentation of plugin jars and dependencies.
+
+Due to that Instrumentation and bytecode upgrades is split into two big parts:
+1. Gradle distribution build time: Interceptor code generation and Gradle API type metadata collection
+2. Gradle runtime: Actual bytecode instrumentation
+
+JavaCompileInterceptorsDeclaration is defined by Gradle developer and with help of annotation processor Gradle then generates bytecode replacements. 
+The actual replacement happens when the user runs his Gradle build.
+
+**_Note: Groovy interception uses the same mechanism but replacements are a bit more complicated due to the dynamic nature of Groovy. 
+And due to that build logic is also intercepted when code is run and not just on the bytecode level._**
+
+Next sections will explain more in detail how we can define interceptors and how the code is then replaced.
+
+
+### Gradle Distribution Build Time
+
+At Gradle Distribution Build time we have two major parts:
+1. Declaring interceptors and generating interceptors
+2. Collecting Gradle API type hierarchy
+
+The first one is important for both configuration cache instrumentation and API upgrades while the second one is used just for API upgrades. 
+Collecting the Gradle API type hierarchy is fully automatic, while declaring interceptors have to be done by Gradle developers. 
+
+```mermaid
+flowchart TB
+    api[[Gradle API]] --> transform{{"Gradle API type\nhierarchy collector\n(Artifact transform)"}}
+    declaredInterceptors[[Interceptors declarations]] --> processor{{"Interceptor generator\n(Annotation processor)"}}
+    processor --> generatedInterceptors[[Generated Interceptors]]
+    transform --> metadata[[Gradle API type metadata]]
+    metadata --> distribution(Gradle distribution)
+    implementation[[Gradle implementation]] --> distribution
+    generatedInterceptors --> distribution
+```
+
+#### Declaring interceptors and generating interceptors
+
+##### Annotation processor
+
+Annotation processor is implemented as Java annotation processor in [platforms/core-runtime/internal-instrumentation-processor](https://github.com/gradle/gradle/tree/master/platforms/core-runtime/internal-instrumentation-processor) and its API (annotations) are defined in the [platforms/core-runtime/internal-instrumentation-api](https://github.com/gradle/gradle/tree/master/platforms/core-runtime/internal-instrumentation-api). 
+This annotation processor transforms interceptor declarations defined by Gradle developers to interceptors that modify Jvm bytecode.
+
+Since this is a Java annotation processor all declarations have to be defined as Java classes, so interceptor declarations in Kotlin or Groovy are not supported.
```

Comment:
I think this is important enough to call a bit more attention to it.

```suggestion
**Note:** Since this is a Java annotation processor all declarations have to be defined as Java classes, so interceptor declarations in Kotlin or Groovy are not supported.
```